### PR TITLE
[CI] Benchmark calling the probe

### PR DESCRIPTION
**Pull request by gukoff**
- Project with 6 benchmarks for firing the probe **while not monitored**
- PR check that compares performance to the `main` branch (benchmark numbers [seem to spread +-25%](https://github.com/benchmark-action/github-action-benchmark?tab=readme-ov-file#stability-of-virtual-environment) on the currently used non-dedicated CI agents)
- Only NativeAOT 8.0 performance is measured. Regular .NET [was seen to be slower](https://github.com/gukoff/DynamicProbes/pull/7#issuecomment-2325205264) during testing.
- The benchmark can be run manually `dotnet run --exporters json --filter '*' -c Release`. I can also add a `Makefile` to simplify this command to `make bench`.
- This action must run [on the default branch](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/caching-dependencies-to-speed-up-workflows#matching-a-cache-key) in this repo at least once in order to cache the "reference" benchmark results and to enable benchmark comparisons for all PRs.

Next steps: 
- measure performance while probed by `bpftrace`. It must be significantly slower due to the 2 context switches from userspace to kernel and back.
- benchmark against the builtin tracing tooling in dotnet.

### CODE REVIEW

File: src/Libstapsdt.cs [hunk 1]
```diff
@@ -6,11 +6,14 @@
 #pragma warning disable CA5392 // Use DefaultDllImportSearchPaths attribute for P/Invokes
                                // ...but it has no effect on non-Windows platforms or the Mono runtime.
 
-namespace Libstapsdt;
+#pragma warning disable CA1401 // P/Invokes should not be visible
```
Comment (gukoff) [inline src/Libstapsdt.cs:9]: Adding a new project to the solution triggered multiple new violations, hence the changes in this file.

(same behaviour persists with `8.0.401` SDK)

File: .github/workflows/benchmark.yaml [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+      # Download previous benchmark result from cache (if exists)
+      - name: Download previous benchmark data
+        uses: actions/cache@v4
+        with:
+          path: ./cache
+          key: last-benchmark
+
+      # This step does multiple things:
+      # - compares these benchmark results with the latest benchmark in the main branch;
+      # - prints the comparison to the CI run sumary
```
Comment (atifaziz) [inline .github/workflows/benchmark.yaml]: nit: typo:

```suggestion
      # - prints the comparison to the CI run summary
```

File: .github/workflows/benchmark.yaml [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+      # - posts the comparison as a PR comment. If a comments already exists, updates it.
```
Comment (atifaziz) [inline .github/workflows/benchmark.yaml]: nit: grammar/typo:

```suggestion
      # - posts the comparison as a PR comment, but updates the comment if it already exists.
```

File: benchmark/Program.cs [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    nint probe3;
+    nint probe4;
+    nint probe5;
+    nint probe6;
+    readonly long arg1 = 1234567890123456789;
+    readonly long arg2 = 2234567890123456789;
+    readonly long arg3 = 3234567890123456789;
+    readonly long arg4 = 4234567890123456789;
+    readonly long arg5 = 5234567890123456789;
+    readonly long arg6 = 6234567890123456789;
```
Comment (atifaziz) [inline benchmark/Program.cs]: These can be `const` fields, no? Any reason to have them as read-only instance fields?

I would also recommend moving them closer to benchmark methods where they are used, like line 47.

File: benchmark/Program.cs [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+using BenchmarkDotNet.Jobs;
+using BenchmarkDotNet.Running;
+using LibstapsdtPinvokes;
+
+namespace Benchmark;
+
+[JsonExporterAttribute.Full]
+[JsonExporterAttribute.FullCompressed]
+[SimpleJob(RuntimeMoniker.NativeAot80)]
+public class NotObservedProbeFireBenchmarks
```
Comment (atifaziz) [inline benchmark/Program.cs]: nit: not + observed = unobserved

```suggestion
public class UnobservedProbeFireBenchmarks
```

File: src/Libstapsdt.cs (hunk 1 above)
Comment (atifaziz) [reply src/Libstapsdt.cs:9]: I think we'll need to make a decision whether the native parts should be directly exposed or not, but this is something we can address later. For now, it's a fair call to make adjustments as needed by flagged items.

**Review by atifaziz — APPROVED** [review]
This is a great addition, thanks!

It would be interesting (at some point in the future) to see how this compares to [other `libstapsdt` wrappers](https://github.com/linux-usdt/libstapsdt/tree/0d53f987b0787362fd9c16a93cdad2c273d809fc?tab=readme-ov-file#wrappers), like Go.

These benchmarks are useful to have, especially to see what would be the effect of adding a higher-level and safer API if there's a plan to do that.

> - The benchmark can be run manually `dotnet run --exporters json --filter '*' -c Release`. I can also add a `Makefile` to simplify this command to `make bench`.

Rather than require `make`, another option is to add a launch profile that `dotnet run` will pick up automatically:

```diff
diff --git a/benchmark/Properties/launchSettings.json b/benchmark/Properties/launchSettings.json
new file mode 100644
index 0000000..4cae0e5
--- /dev/null
+++ b/benchmark/Properties/launchSettings.json
@@ -0,0 +1,8 @@
+{
+  "profiles": {
+    "Benchmark": {
+      "commandName": "Project",
+      "commandLineArgs": "--exporters json --filter *"
+    }
+  }
+}
```

And then all you need is this:

    dotnet run --project benchmark -c Release

This is something that the workflow can also benefit from and the above CLI could even go into a simple `bench` shell script. Yet another option is to put all of it into the shell script:

```sh
set -e
cd "$(dirname "$0")"
dotnet run --project benchmark -c Release -- --exporters json --filter '*'
```

> Only NativeAOT 8.0 performance is measured. Regular .NET [was seen to be slower](https://github.com/gukoff/DynamicProbes/pull/7#issuecomment-2325205264) during testing.

Why not publish both numbers?

File: benchmark/Program.cs (hunk 4 above)
Comment (gukoff) [reply benchmark/Program.cs]: Yup, no reason to keep them readonly, changed to const.

**Comment by gukoff** [comment]
>> Only NativeAOT 8.0 performance is measured. Regular .NET [was seen to be slower](https://github.com/gukoff/DynamicProbes/pull/7#issuecomment-2325205264) during testing.

> Why not publish both numbers?

The github action that publishes the performance diff with the main branch does not differentiate between the [RuntimeMonikers](https://benchmarkdotnet.org/articles/overview.html#jobs). It would show two rows in the table with the test name `Benchmark.UnobservedProbeFireBenchmarks.ArgCount2`, and you wouldn't know which one is NativeAOT.

The table I linked above is one of the artifacts generated when running the benchmark. It has more metadata for the benchmark results, but it does not compare performance to the main branch.

Likely these issues can be resolved by a finer configuration of the pipeline, which I'll look into next.

**Pull request merged**